File: CPP/WebDev/WebDev/WebFrameworkLib/Request.cpp
#include "WebFramework/Requests/Request.h"

#include <istream>
#include <sstream>
#include <fstream>
#include <vector>
#include <iterator>

#include <iostream>

void Request::parse() {
	std::istringstream iss(m_request);
	std::vector<std::string> parsed((std::istream_iterator<std::string>(iss)), std::istream_iterator<std::string>());

	method = parsed[0];
	route = parsed[1];

	std::string route_params = "";
	int idx = route.find('?');

	if (idx != std::string::npos) {
		route_params = route.substr(idx + 1);
		route = route.substr(0, idx);
		params = WebServerUtil::parse_attribute_string(route_params);
	}

	protocol = parsed[3];
	host = parsed[7];

	std::string post_data = parsed[parsed.size() - 1];
	
	if (method == "POST") {
		data = WebServerUtil::parse_attribute_string(post_data);
	}
}

void Request::setContent(std::string content) {
	output_content = content;
	output_code = 200;
}

void Request::readFile(std::string filePath) {
	output_content = WebServerUtil::FileParser::parseFile(filePath, this);
}

void Request::forward() {
	// write the document back to the client
	std::ostringstream oss;
	oss << "HTTP/1.1 " << output_code << " OK\r\n";
	oss << "Cache-Control: no-cache, private\r\n";
	oss << "Content-Type: text/html\r\n";
	oss << "Content-Length: " << output_content.size() << "\r\n";
	oss << "\r\n";
	oss << output_content;

	std::string output = oss.str();
	int size = output.size() + 1;

	send(sendingClient.socket, output.c_str(), size, 0);
}

void Request::setContext(std::string key, std::string value) {
	sendingClient.context[key] = value;
}

std::map<std::string, std::string> Request::context() {
	return sendingClient.context;
}

File: Python/WebDev/WebServer/sockets/Request.py
from jinja2 import Template

def parseAttributeString(atts):
    data = {}
    for param in atts.split("&"):
        sep_idx = param.find("=")
        key = param[:sep_idx]
        value = param[sep_idx + 1:]
        data[key] = value

    return data

class Request:
    """
    std::string							output_content;		// content to be sent back
        int									output_code;		// output code
        std::string							route;				// request route
        std::string							method;				// method of request
        std::string							protocol;			// request protocol
        std::string							host;				// request host
        std::map<std::string, std::string>	data;				// request form data
        std::map<std::string, std::string>	params;				// specified parameters in url request
    """

    def __init__(self, request, client):
        self.request_body = request
        self.client = client
        self.method = ""
        self.route = ""
        self.params = {}
        self.data = {}
        self.response_content = ""
        self.response_code = 200
        self.response = ""
        self.type = "text/html"

    def parse(self):
        request = self.request_body.split()

        # request method
        self.method = request[0]

        route = request[1]
        self.route = route

        if self.route.find(".css") == -1:
            query_idx = route.find("?")
            if query_idx != -1:
                # request parameters'
                self.route = route[:query_idx]
                params = route[query_idx + 1:]
                self.params = parseAttributeString(params)

            # form data
            data = request[-1]
            if data.find("=") != -1:
                # has data parameters from form
                self.data = parseAttributeString(data)
        else:
            self.type = "text/css"
            self.render_template(self.route[1:])

    def render_template(self, file_path):
        content = ""
        with open(file_path, 'r') as content_file:
            content = content_file.read()

        if self.route.find(".css") == -1:
            self.render_content(content)
        else:
            self.response_content = content

    def render_content(self, content):
        template = Template(content)
        self.response_content = template.render(self.client.context)

    def getResponse(self):
        return_char = "\r\n"
        self.response = "HTTP/1.1 " + str(self.response_code) + " OK" + return_char + \
                        "Cache-Control: no-cache, private" + return_char + \
                        "Content-Type: " + self.type + return_char + \
                        "Content-Length: " + str(len(self.response_content)) + return_char + \
                        str(return_char) + \
                        self.response_content

        return self.response